### DISCUSSION TITLE
Subscriptions help

### DISCUSSION THREAD

**Discussion by thuperthecret2** [hayes/pothos#732] [Q&A] [answered]
4 years ago I wrote a full stack JS app with subscriptions using Apollo and Prisma1, and PostgresPubSub from graphql-postgres-subscriptions 1.0.5.
tl;dr I'm trying to get subscriptions working in a new server and app in TS using Pothos, Apollo Server 4, and Prisma 4.8.  I've got the server running and all the CRUD working but I've been trying for a couple days on subscriptions and having not much luck.  On a side note I don't think I can use PostgresPubSub anymore as it likely won't work with graphql 16.6, so I'll just use PubSub from 'graphql-subscriptions' 2.0.0.

For example I need subscriptions working on a User entity (let's assume User just has 'id' and 'name' for now).  I need a subscription only for a certain user.id that had a mutation type of UPDATED (not CREATED or DELETED) and to get back only the info I ask for (in this case 'id' and 'name').

The previous autogenerated server graphql looked like this 4 years ago:

```
type UserSubscriptionPayload {
  mutation: MutationType!
  node: User
  updatedFields: [String!]
  previousValues: UserPreviousValues
}

input UserSubscriptionWhereInput {
  mutation_in: [MutationType!]
  updatedFields_contains: String
  updatedFields_contains_every: [String!]
  updatedFields_contains_some: [String!]
  node: UserWhereInput
  AND: [UserSubscriptionWhereInput!]
  OR: [UserSubscriptionWhereInput!]
  NOT: [UserSubscriptionWhereInput!]
}
```

The Apollo client gql (on the web client) looked like this:

```
subscription UserUpdatedSubscription($id: ID!) {
  user(where: { mutation_in: [UPDATED], node: { id: $id } }) {
    mutation
    node {
      id
      name
    }
  }
}
```

I know this is a broad question, but is this possible using Pothos SmartSubscriptions, and are there any examples of it?  If it's not possible with SmartSubscriptions, would I be able to use something like builder.subscriptionField?

Here's my attempt at SmartSubscriptions:
```
builder.queryField('user', (t) =>
  t.prismaField({
    type: 'User',
    nullable: true,
    args: {
      id: t.arg.string({ required: true }),
    },
    smartSubscription: true,
    subscribe: (subscriptions, root, args, ctx, info) => {
      subscriptions.register('user')
      console.log('Subscription registered on user.id: ' + args.id)
    },
    resolve: async (query, root, args, ctx, info) => {
      return prisma.user.findUnique({
        ...query,
        where: {
          id: args.id,
        },
      })
    },
  }),
)
```
which results in the following auto-generated in the schema.graphql:
```
type Subscription {
  user(id: String!): User
}
```
so I don't quite understand how to get the mutationType out of that.

If I can't do any or all of this using Pothos, how do I go about adding subscriptions to my graphql schema that are _apart_ from Pothos-generated code?  I will need file uploads and authentication next and not everything can be configured in Pothos the way I might specifically need it.

**Comment by hayes** [comment, marked as answer]
Added an example here, hope this helps: https://github.com/hayes/pothos/tree/main/examples/prisma-subscriptions

The short answer is that prisma doesn't support automatically publishing events when things change, so you will need to manually "publish" events when you mutate data.  The exact format for how you handle this is up to you, but hopefully the example above will help

**Reply by thuperthecret2** [reply]
Thanks, much appreciated!  I will take a look and let you know the results!

**Reply by thuperthecret2** [reply]
It worked great!  I switched to graphql-yoga (instead of Apollo) to get it working.  Maybe I'll stick with yoga.
Here's what the `userSubscription` initially returned on an `updateUser` mutation:
```
{
  "data": {
    "userSubscription": {
      "user": {
        "name": "Test5"
      },
      "mutationType": "UPDATED"
    }
  }
}
```
